controllers/bmcjob_controller.go

```diff
@@ -18,38 +18,207 @@ package controllers
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Runs each Task in a BMCTask and updates the BMCJob status.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BMCJob", req.NamespacedName)
+	logger.Info("Reconciling BMCJob")
+
+	// Fetch the bmcJob object
+	bmcJob := &bmcv1alpha1.BMCJob{}
+	err := r.client.Get(ctx, req.NamespacedName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job has StartTime is noop.
+	if !bmcJob.Status.StartTime.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	// Create a patch from the initial BMCJob object
+	// Patch is used to update Status after reconciliation
+	bmcJobPatch := client.MergeFrom(bmcJob.DeepCopy())
+
+	return r.reconcile(ctx, bmcJob, bmcJobPatch, logger)
+}
+
+func (r *BMCJobReconciler) reconcile(ctx context.Context, bmj *bmcv1alpha1.BMCJob, bmjPatch client.Patch, logger logr.Logger) (ctrl.Result, error) {
+	// Get BaseboardManagement object for the Job
+	// Requeue if error
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.resolveBaseboardManagementRef(ctx, bmj, baseboardManagement)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BMCJob %s/%s BaseboardManagementRef: %v", bmj.Namespace, bmj.Name, err)
+	}
+
+	// Fetching username, password from SecretReference
+	// Requeue if error fetching secret
+	username, password, err := resolveAuthSecretRef(ctx, r.client, baseboardManagement.Spec.Connection.AuthSecretRef)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BaseboardManagement %s/%s SecretReference: %v", baseboardManagement.Namespace, baseboardManagement.Name, err)
+	}
+
+	// Initializing BMC Client
+	bmcClient, err := r.bmcClientFactory(ctx, baseboardManagement.Spec.Connection.Host, strconv.Itoa(baseboardManagement.Spec.Connection.Port), username, password)
+	if err != nil {
+		logger.Error(err, "BMC connection failed", "host", baseboardManagement.Spec.Connection.Host)
+		bmj.SetCondition(bmcv1alpha1.JobFailed, bmcv1alpha1.BMCJobConditionTrue, bmcv1alpha1.WithJobConditionMessage(fmt.Sprintf("Failed to connect to BMC: %v", err)))
+		result, patchErr := r.patchBMCJobStatus(ctx, bmj, bmjPatch)
+		if patchErr != nil {
+			return result, utilerrors.NewAggregate([]error{patchErr, err})
+		}
+
+		return result, err
+	}
+
+	// Pause BaseboardManagement reconciliation before running BMCJob tasks
+	baseboardManagement.PauseReconcile()
```

Comment:
2 problems here. 

(1) This doesn't pause anything; it just sets a puase annotation on the local copy of the `BaseboardManagement` type. I.e. The etcd store still has the unpaused type so reconciliation is still on-going. 

(2) I appreciate there could be a conflict between the `BaseboardManagement` power reconciliation and a power task (as well as any future `BaseboardManagement` reconciliation that may be added) but I don't think a job submission should control whether or not they collide. The operator needs to opt-in to that so pausing should probably be a thing an operator specifies at the task level. If we find its not paused, we can log some information specifying that so its clear. Let me know what you think.